README.md

```diff
@@ -0,0 +1,71 @@
+[![Build Status](https://github.com/gafter/AutoHashEqualsCached.jl/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/gafter/AutoHashEqualsCached.jl/actions/workflows/CI.yml?query=branch%3Amain)
+
+# AutoHashEqualsCached
+
+A pair of macros to add `==` and `hash()` to struct types: `@auto_hash_equals` and `@auto_hash_equals_cached`.
+
+# `@auto_hash_equals`
+
+The macro `@auto_hash_equals` produces an implementation of `Base.hash(x)` that computes the hash code when invoked.  It is offered for compatibility with the package [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl).
+
+You use it like so:
+
+```julia
+@auto_hash_equals struct Box
+    x
+end
+```
+
+which is translated to
+
+```julia
+struct Box
+    x
+end
+Base.hash(x::Type, h::UInt) = hash(x.x, hash(:Foo, h))
+Base.(:(==))(a::Box, b::Box) = isequal(a.x, b.x)
+```
+
+However, there are a few enhancements beyond what is provided by [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl):
+
+- It handles empty structs without error
+- You can use it either before or after another macro.
+
+For compatibility with [AutoHashEquals](https://github.com/andrewcooke/AutoHashEquals.jl), we do not take the type arguments of a generic type into account in `==`.  So a `Box{Int}(1)` will test equal to a `Box{Any}(1)`.
+
+# `@auto_hash_equals_cached`
+
+The macro `@auto_hash_equals_cached` is useful for non-mutable struct types that define recursive or deep data structures (and therefore are likely to be stored on the heap).  It computes the hash code during construction and caches it in a field of the struct.  If you are working with data structures of any significant depth, computing the hash once can speed things up at the expense of one additional field per struct.
+
+You use it like so:
+
+```julia
+@auto_hash_equals_cached struct Point{T<:Any, U<:Any}
+  x::T
+  y::U
+end
+```
+
+which is translated to
+
+```julia
+struct Point{T<:Any, U<:Any}
+    x::T
+    y::U
+    _cached_hash::UInt
+    Point{T,U}(x::T, y::U) where {T<:Any, U<:Any} = new{T,U}(x,y,hash(y, hash(x, hash(Point{T,U}))))
```

Comment:
This is actually slightly more restrictive than the default constructor that's generated by Julia.

I wish there was an easy way to see this via reflection; something like `@code_struct struct Foo ... end`, but there isn't. (I filed an issue for that a year ago, here: https://github.com/JuliaLang/julia/issues/44943)

But from what I understand, the default constructors for a type like `Point` would be essentially:
```julia
struct Point{T<:Any, U<:Any}
  x::T
  y::U
  Point{T,U}(x, y) where {T,U} = new{T,U}(convert(T, x), convert(U, y))
  Point(x::T, y::U) where {T,U} = new{T,U}(x, y)
end
```
That is: there are two ways this are less restrictive than yours:
1. They provide a constructor that infers the types: `Point(1, 2.0) -> Point{Int64, Float64}(1, 2.0)`
2. They let you coerce values into declared types: `Point{Int,Int}(1, 2.0) -> Point{Int64, Int64}(1, 2)`

Maybe we could change the generated constructors to be something similar? I.e.:
```julia
struct Point{T<:Any, U<:Any}
    x::T
    y::U
    _cached_hash::UInt
    # infer the types
    Point(x::T, y::U) where {T, U} = new{T,U}(x, y, hash(y, hash(x, hash(Point{T,U}))))
    # allow conversion:
    function Point{T,U}(xx, yy) where {T, U}
        x, y = convert(T, xx), convert(U, yy)
        new{T,U}(x, y, hash(y, hash(x, hash(Point{T,U}))))
    end
end
```
I think this would behave much more like a normal julia struct declaration. :)

It's hard to tell from your source code below whether you're generating something like this already or not.

Reply:
I created issue #12 to address this